packages/react-reconciler/src/ReactFiberHooks.new.js

```diff
@@ -393,6 +393,13 @@ export function renderWithHooks<Props, SecondArg>(
       current !== null && current.type !== workInProgress.type;
   }
 
+  const memoCache = workInProgress.memoCache;
+  if (memoCache !== null) {
+    // Clone memo cache prior to rendering to avoid corruption in case of error
+    memoCache.previous = memoCache.current;
+    memoCache.current = [...memoCache.current];
+  }
+
```

Comment:
I think there is an extra layer of indirection that is not needed here. If memoCache wasn't an object with prev/current properties but was instead the cache itself then when `workInProgress` is created it will get a pointer to the existing cache and then in `renderWithHooks` you can copy in place.

```
if (memoCache !== null) {
  workInProgress.memoCache = [...memoCache];
}
```

This maybe better embeds the meaning of previous / current within the Fiber tree structure since each generation of memoCache lives the same lifetime as it's host Fiber. The current setup requires managing the migration from current to previous which simple enough is seemingly extraneous from what I can gather.

Reply:
Yeah, I wasn't sure if it would be sufficient to rely on the alternate fiber as the "previous" value, if both the current/alternate version of a fiber could be concurrently updating. If that can't happen, then yeah the indirection here is unnecessary.